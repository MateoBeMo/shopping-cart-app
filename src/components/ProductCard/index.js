import ProductCart from './ProductCard';

export default ProductCart;
