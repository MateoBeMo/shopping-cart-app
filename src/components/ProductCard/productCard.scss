@import 'styles/styles-utils.scss';

.productCard__wrapper {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 300px;
    height: 300px;
    margin: auto;
    border: 1px solid #999;
    border-radius: 0.3em;
    box-shadow: 0 3px 3px 0 rgba(0, 0, 0, 0.16), 0 3px 4px 0 rgba(0, 0, 0, 0.18), 0 1px 8px 0 rgba(0, 0, 0, 0.2);
    .productCard__header {
        position: relative;
        width: 100%;
        min-height: 180px;
        .productCard__image {
            width: 100%;
            height: 180px;
            border-bottom: 1px solid #999;
            border-top-left-radius: 0.3em;
            border-top-right-radius: 0.3em;
            object-fit: cover;
        }
        .productCard__favorite {
            cursor: pointer;
            position: absolute;
            left: 270px;
            top: 15px;
            user-select: none;
            &:hover {
                transform: scale(1.3);
            }
            &--red {
                svg {
                    path {
                        fill: red;
                    }
                }
            }
        }
    }
    .productCard__content {
        display: flex;
        flex-direction: column;
        justify-content: space-between;
        height: 100%;
        padding: 8px 16px;
        .productCard__title {
            display: flex;
            justify-content: space-between;
            & span:last-child {
                font-weight: 525;
            }
        }
        .productCard__description {
            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
            width: 280px;
        }
        .productCard__footer {
            display: flex;
            justify-content: space-between;
            span {
                color: gray;
            }
            .productCard__button {
                cursor: pointer;
                padding: 8px 10px;
                border-radius: 1.5em;
                &--disabled {
                    border: 1px solid #999999;
                    background-color: #cccccc;
                    color: #666666;
                    cursor: initial;
                }
            }
        }
    }
}

.skeleton__wrapper {
    border: none;
    box-shadow: none;
}

.skeleton__line {
    height: 100%;
    width: 100%;
    background: linear-gradient(-90deg, #ddd 0%, #eee 50%, #ddd 100%);
    background-size: 400% 400%;
    animation: pulse 1.2s ease-in-out infinite;
    @keyframes pulse {
        0% {
            background-position: 0% 0%;
        }
        100% {
            background-position: -135% 0%;
        }
    }
    // width: 5.5em;
    border-radius: 5px;
    &::before {
        content: '\00a0';
    }
}
