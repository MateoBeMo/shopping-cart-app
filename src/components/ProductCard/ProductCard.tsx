import React from 'react';
/**Libraries */
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import classNames from 'classnames';
/**Components */
import Button from 'components/Button';
/**Styles */
import './productCard.scss';

type ProductCardProps = {
    image_url: string;
    stock: number;
    productName: string;
    price: number;
    productDescription: string;
    favorite: string;
    onSelected: (e: React.MouseEvent) => void;
    toggleFavorite: (e: React.MouseEvent) => void;
};

const ProductCard = ({
    image_url,
    stock,
    productName,
    price,
    productDescription,
    favorite,
    onSelected,
    toggleFavorite,
}: ProductCardProps) => {

    const favoriteClass = classNames('productCard__favorite', { "productCard__favorite--red": favorite === "1" });
    const addProductClass = classNames('productCard__button', { "productCard__button--disabled": stock === 0 });

    return (
        <div className="productCard productCard__wrapper">
            <div className="productCard__header">
                <img className="productCard__image" src={image_url} alt={productName} />
                <span className={favoriteClass} onClick={toggleFavorite} data-testid="heart-icon">
                    <FontAwesomeIcon icon="heart" />
                </span>
            </div>
            <div className="productCard__content">
                <div className="productCard__title">
                    <span>{productName}</span>
                    <span>{price}$</span>
                </div>
                <div className="productCard__description" title={productDescription}>{productDescription}</div>
                <div className="productCard__footer">
                    <span>{stock} left</span>
                    <Button className={addProductClass} onClick={onSelected} disabled={stock === 0}>
                        <FontAwesomeIcon icon="plus" />
                    </Button>
                </div>
            </div>
        </div>
    );
};

export default ProductCard;
