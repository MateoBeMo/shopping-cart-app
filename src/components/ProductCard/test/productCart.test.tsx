import React from 'react';
/**Components */
import ProductCard from '../ProductCard';
/**Testing library */
import { render, fireEvent, cleanup } from '@testing-library/react';
import { axe } from 'jest-axe';


type ProductCardProps = {
    image_url: string;
    stock: number;
    productName: string;
    price: number;
    productDescription: string;
    favorite: string;
    onSelected: (e: React.MouseEvent) => void;
    toggleFavorite: (e: React.MouseEvent) => void;
};

const onSelectedMock = jest.fn();
const toggleFavoriteMock = jest.fn();


const renderProductCard = (props: Partial<ProductCardProps> = {}) => {
  const defaultProps: ProductCardProps = {
    onSelected:  onSelectedMock,
    toggleFavorite: toggleFavoriteMock,
    // eslint-disable-next-line @typescript-eslint/camelcase
    image_url: 'https://dummyimage.com/400x400/bc27b5/000&text=Gorgeous Frozen Chair',
    stock: 28,
    productName: 'Gorgeous Frozen Chair',
    price: 47,
    productDescription:
        'Harum modi sunt. Voluptatem ut molestiae consequatur. Ea omnis architecto laboriosam accusantium reiciendis corporis exercitationem ad dolor. Fugit autem placeat voluptas sint aut aliquam sed. Totam fuga nesciunt rerum voluptatibus. Voluptatibus voluptates vel ut et temporibus perferendis laboriosam accusamus.',
    favorite: '1',
  };
  return render(<ProductCard {...defaultProps} {...props} />);
}

describe('<ProductCard>', () => {
    afterEach(() => {
        cleanup();
    });

    test('Snapshot renders => it should rendered with the provided props', () => {
        const { container } = renderProductCard();

        expect(container).toMatchSnapshot();
    });

    test('It should fire toggleFavorite event on click over the heart icon', () => {
        const { getByTestId } = renderProductCard()

        const heartIcon = getByTestId('heart-icon');

        fireEvent.click(heartIcon);

        expect(toggleFavoriteMock).toHaveBeenCalled();
        expect(toggleFavoriteMock).toHaveBeenCalledTimes(1);
    });

    test('It should fire onSelected event on click over add product button', async() => {
        const { findByText } = renderProductCard()

        const addProductButton = await findByText('Add');

        fireEvent.click(addProductButton);

        expect(onSelectedMock).toHaveBeenCalled();
        expect(onSelectedMock).toHaveBeenCalledTimes(1);
    });

    test('should not have basic accessibility issues', async () => {
        const { container } = renderProductCard();
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
});
