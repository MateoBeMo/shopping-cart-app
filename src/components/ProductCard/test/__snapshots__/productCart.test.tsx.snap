// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard> Snapshot renders => it should rendered with the provided props 1`] = `
<div>
  <div
    class="productCard productCard__wrapper"
  >
    <div
      class="productCard__header"
    >
      <img
        alt="Gorgeous Frozen Chair"
        class="productCard__image"
        src="https://dummyimage.com/400x400/bc27b5/000&text=Gorgeous Frozen Chair"
      />
      <span
        class="productCard__favorite productCard__favorite--red"
        data-testid="heart-icon"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-heart fa-w-16 "
          data-icon="heart"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
    <div
      class="productCard__content"
    >
      <div
        class="productCard__title"
      >
        <span>
          Gorgeous Frozen Chair
        </span>
        <span>
          47
          $
        </span>
      </div>
      <div
        class="productCard__description"
      >
        Harum modi sunt. Voluptatem ut molestiae consequatur. Ea omnis architecto laboriosam accusantium reiciendis corporis exercitationem ad dolor. Fugit autem placeat voluptas sint aut aliquam sed. Totam fuga nesciunt rerum voluptatibus. Voluptatibus voluptates vel ut et temporibus perferendis laboriosam accusamus.
      </div>
      <div
        class="productCard__footer"
      >
        <span>
          28
           left
        </span>
        <button
          class="productCard__button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-plus fa-w-14 "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              fill="currentColor"
            />
          </svg>
           Add
        </button>
      </div>
    </div>
  </div>
</div>
`;
